presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,793 @@
+
+            int offset = 0;
+            do {
+                int batchSize = calculateNextBatchSize(fixedWidthRowSize, variableWidthChannels, offset);
+
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i].setNextBatch(offset, batchSize);
+                    blockEncodingBuffers[i].copyValues();
+                }
+
+                bufferedRowCount += batchSize;
+                offset += batchSize;
+
+                if (bufferFull) {
+                    flush(outputBuffer);
+
+                    bufferFull = false;
+                }
+            }
+            while (offset < positionCount);
+        }
+
+        public long getSizeInBytes()
+        {
+            long buffersSizeInBytes = 0;
+            if (blockEncodingBuffers != null) {
+                for (int i = 0; i < channelCount; i++) {
+                    buffersSizeInBytes += blockEncodingBuffers[i].getSizeInBytes();
+                }
+            }
+
+            // Add up sizes for positions and rowSizes arrays
+            return SIZE_OF_INT * 2 * positionCount + buffersSizeInBytes;
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            long buffersSizeInBytes = 0;
+            if (blockEncodingBuffers != null) {
+                for (int i = 0; i < channelCount; i++) {
+                    buffersSizeInBytes += blockEncodingBuffers[i].getRetainedSizeInBytes();
+                }
+            }
+
+            return INSTANCE_SIZE + sizeOf(positions) + sizeOf(rowSizes) + buffersSizeInBytes;
+        }
+
+        private void initializeBlockEncodingBuffers(DecodedBlockNode[] decodedBlocks)
+        {
+            // Create buffers has to be done after seeing the first page.
+            if (blockEncodingBuffers == null) {
+                int initialElementCount = min(capacity / channelCount / DEFAULT_ELEMENT_SIZE_IN_BYTES, INITIAL_POSITION_COUNT);
+
+                blockEncodingBuffers = new BlockEncodingBuffers[channelCount];
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i] = createBlockEncodingBuffers(decodedBlocks[i], initialElementCount);
+                }
+            }
+        }
+
+        private void calculateRowSizes(int fixedWidthRowSize, List<Integer> variableWidthChannels)
+        {
+            if (variableWidthChannels.isEmpty()) {
+                return;
+            }
+
+            ensureRowSizesCapacityAndInitialize();
+
+            for (int i : variableWidthChannels) {
+                blockEncodingBuffers[i].accumulateRowSizes(rowSizes);
+            }
+
+            for (int i = 0; i < positionCount; i++) {
+                rowSizes[i] += fixedWidthRowSize;
+            }
+        }
+
+        private int calculateNextBatchSize(int fixedWidthRowSize, List<Integer> variableWidthChannels, int startPosition)
+        {
+            int bytesRemaining = capacity - toIntExact(getSerializedBuffersSizeInBytes());
```

Comment:
instead of calling `toIntExact`, change the return type of `getSerializedBuffersSizeInBytes` to `int`